src/vm/database/marf.rs

```diff
@@ -14,83 +18,161 @@ use chainstate::burn::BlockHeaderHash;
+pub struct MarfedKV {
     chain_tip: BlockHeaderHash,
     marf: MARF,
     // Since the MARF only stores 32 bytes of value,
     //   we need another storage
-    side_store: S
+    side_store: SqliteConnection
 }
 
-#[cfg(test)]
-pub fn temporary_marf() -> MarfedKV<std::collections::HashMap<String, String>> {
-    use std::env;
-    use rand::Rng;
-    use util::hash::to_hex;
-    use std::collections::HashMap;
+pub struct MemoryBackingStore {
+    side_store: SqliteConnection
+}
 
-    let mut path = env::temp_dir();
-    let random_bytes = rand::thread_rng().gen::<[u8; 32]>();
-    path.push(to_hex(&random_bytes));
+// These functions generally _do not_ return errors, rather, any errors in the underlying storage
+//    will _panic_. The rationale for this is that under no condition should the interpreter
+//    attempt to continue processing in the event of an unexpected storage error.
+pub trait ClarityBackingStore {
+    /// put K-V data into the committed datastore
+    fn put_all(&mut self, items: Vec<(String, String)>);
+    /// fetch K-V out of the committed datastore
+    fn get(&mut self, key: &str) -> Option<String>;
+    fn has_entry(&mut self, key: &str) -> bool {
+        self.get(key).is_some()
+    }
 
-    let marf = MARF::from_path(path.to_str().expect("Inexplicably non-UTF-8 character in filename"), None)
-        .unwrap();
-    let side_store = HashMap::new();
+    /// change the current MARF context to service reads from a different chain_tip
+    ///   used to implement time-shifted evaluation.
+    /// returns the previous block header hash on success
+    fn set_block_hash(&mut self, bhh: BlockHeaderHash) -> Result<BlockHeaderHash>;
 
-    let chain_tip = TrieFileStorage::block_sentinel();
+    fn get_block_at_height(&mut self, height: u32) -> Option<BlockHeaderHash>;
+    fn get_open_chain_tip_height(&mut self) -> u32;
+    fn get_open_chain_tip(&mut self) -> BlockHeaderHash;
+    fn get_side_store(&mut self) -> &mut SqliteConnection;
 
-    MarfedKV { chain_tip, marf, side_store }
-}
+    /// The contract commitment is the hash of the contract, plus the block height in
+    ///   which the contract was initialized.
+    fn make_contract_commitment(&mut self, contract_hash: Sha512Trunc256Sum) -> String {
+        let block_height = self.get_open_chain_tip_height();
+        let cc = ContractCommitment { hash: contract_hash, block_height };
+        cc.serialize()
+    }
 
-#[cfg(test)]
-pub fn in_memory_marf() -> MarfedKV<SqliteConnection> {
-    use std::env;
-    use rand::Rng;
-    use util::hash::to_hex;
-    use std::collections::HashMap;
+    /// This function is used to obtain a committed contract hash, and the block header hash of the block
+    ///   in which the contract was initialized. This data is used to store contract metadata in the side
+    ///   store.
+    fn get_contract_hash(&mut self, contract: &QualifiedContractIdentifier) -> Result<(BlockHeaderHash, String)> {
```

Comment:
Looking at the usage of this method, is it important that it returns a `String` instead of just a bare `Sha512Trunc256Sum`?  Is the hash ever used?

Reply:
Nope, the hash is not used anywhere at the moment -- changed to just return the bare hash struct in e325dca04.